### PR TITLE
HDDS-7098. Provide a way for admin to identify all unhealthy container replicas

### PR DESCRIPTION
**Pull request by mladjan-gadzic**
## What changes were proposed in this pull request?

Add state field for container replica response for Recon API.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7098

## How was this patch tested?

- manual
- unit

### CODE REVIEW

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto
Comment (mladjan-gadzic) [inline hadoop-hdds/interface-client/src/main/proto/hdds.proto:408]: An idea was to pull an enum from https://github.com/mladjan-gadzic/ozone/blob/HDDS-7098/hadoop-hdds/interface-server/src/main/proto/ScmServerDatanodeHeartbeatProtocol.proto#L213 and use it here in order not to maintain same enum in two places but I could not make it. Because of that I decided not to use enum at all and stick with string instead.

**Review by devmadhuu — APPROVED** [review]

**Comment by kerneltime** [comment]
cc @adoroszlai can you please take a look?

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto
Comment (jojochuang) [reply hadoop-hdds/interface-client/src/main/proto/hdds.proto:408]: will it be backward compatible if it's a required field, or is backward compat not a concern here?

Comment (mladjan-gadzic) [reply hadoop-hdds/interface-client/src/main/proto/hdds.proto:408]: I'd say backward compatibility is not a concern here.

Comment (jojochuang) [reply hadoop-hdds/interface-client/src/main/proto/hdds.proto:408]: It would break rolling upgrade, or when recon and SCM are in different versions.

I see that rolling upgrade is not yet a thing, so I am okay with it.

**Review by jojochuang — COMMENTED** [review]
It would be idea to reuse the State Enum type.
Would the answer in this stackoverflow article help address the import problem? https://stackoverflow.com/questions/66524793/how-to-use-the-enum-of-one-proto-file-in-another-file

**Comment by mladjan-gadzic** [comment]
> It would be ideal to reuse the State Enum type. Would the answer in this stackoverflow article help address the import problem? https://stackoverflow.com/questions/66524793/how-to-use-the-enum-of-one-proto-file-in-another-file

I like that idea as well. I've tried it, but there is an issue with backward compatibility: `[ERROR] Failed to execute goal com.salesforce.servicelibs:proto-backwards-compatibility:1.0.7:backwards-compatibility-check (default) on project hdds-interface-server: Backwards compatibility check failed! You can override this by specifying allowBreakingChanges=true -> [Help 1]`. I'd say we should not break it, especially because I am not aware of the consequences it could have on the system

**Comment by adoroszlai** [comment]
> > It would be ideal to reuse the State Enum type. Would the answer in this stackoverflow article help address the import problem? https://stackoverflow.com/questions/66524793/how-to-use-the-enum-of-one-proto-file-in-another-file
> 
> I've tried it, but there is an issue with backward compatibility

Does the compatibility check also fail if the field is `optional` instead of `required`?

**Comment by mladjan-gadzic** [comment]
> > > It would be ideal to reuse the State Enum type. Would the answer in this stackoverflow article help address the import problem? https://stackoverflow.com/questions/66524793/how-to-use-the-enum-of-one-proto-file-in-another-file
> > 
> > 
> > I've tried it, but there is an issue with backward compatibility
> 
> Does the compatibility check also fail if the field is `optional` instead of `required`?

Yes, it does. Backward compatibility check fails for `hdds-interface-server` from which I've tried to pull an enum. Furthermore I tried using `optional` there, but no avail.

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto
Comment (kerneltime) [reply hadoop-hdds/interface-client/src/main/proto/hdds.proto:408]: There needs to be a powerful reason for making the proto message `required`. I would change this to optional. If the code depends on it, it should handle it rather than having proto layer throw an exception on parsing.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java
Comment (errose28) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:444]: Do we need to update the replica state here too?

**Review by errose28 — COMMENTED** [review]
Not being able to use the enum is somewhat unfortunate, but there is a precedent for using the string state in hdds.proto from `SCMContainerReplicaProto`, which likely had to use a string for the same reason. I have one question inline, but overall LGTM if others are satisfied with the current state of the protos.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java
Comment (mladjan-gadzic) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:444]: Thanks for the review! Yes, you are right, I'll add that.

**Review by errose28 — APPROVED** [review]
Thanks for the update @mladjan-gadzic LGTM. Does anyone involved in the earlier proto discussion have outstanding concerns? @kerneltime @jojochuang @adoroszlai

**Comment by mladjan-gadzic** [comment]
@errose28 thanks for the review!

**Comment by adoroszlai** [comment]
Failure of `TestOzoneManagerHASnapshot` is unrelated and it has been fixed on `master` since then.

**Comment by adoroszlai** [comment]
Thanks @mladjan-gadzic for the patch, @devmadhuu, @errose28, @jojochuang, @kerneltime for the review.

**Pull request merged**
